### DISCUSSION TITLE
Issue when running the OpenMage cron in a Docker container (DDEV case)

### DISCUSSION THREAD

**Discussion by addison74** [OpenMage/magento-lts#3795] [Questions & Answers]
Running of the cron.sh script from the OpenMage root directory populates the **cron_schedule** table with data. If it is done automatically, then the following line must be added in the user's crontab
```
* * * * * /var/www/html/cron.sh
```
This means that the script is running every minute.

DDEV makes testing easier and is based on Docker. It configures so that in less than a minute you can start an instance of OpenMage and provides useful commands to import/export the database by creating snapshots, run commands in containers and many more. Changing the configuration you can use immediately Nginx or Apache, PHP any version.

I managed to configure the cron and if I execute the command ```ddev exec crontab -l``` the line with the OpenMage script is displayed, but the table in the database is not populated with data. I waited a long time and nothing happened. I ran the script directly in the container without success.

Without a functional configuration, the OpenMage cron cannot be run, especially the sending of notifications after the order, newsletter subscription, ...

I will try finding a solution to make this important part of the OpenMage test environment based on DDEV work. This is the case for (Windows 10/11 + Mkcert + WSL + Docker + DDEV).

**Comment by pquerner** [comment]
For debugging crons I use n98 on the cli, or if I want to Aoe_Scheduler from the backend. 
Most of the time I don't need crons running in a debuggable environment and only when I am looking for stuff. Hence n98 or Aoe_Scheduler.

So, why is it needed to have a perma cron run in ddev?

**Reply by fballiano** [reply]
I don't use ddev but I agree with this, I never run cron on my dev environment, if needed I trigger them manually. Actually I prefer having them disabled cause sometimes they can do stuff which is only supposed to run in production (send emails and stuff) which I wouldnt want to run on my local computer.

**Reply by addison74** [reply]
This discussion was initiated to solve a problem that I discovered in the use of DDEV. We are not discussing why I want to use a cron in a dev environment or not.

**Reply by fballiano** [reply]
ah ok I didn't read correctly the whole thing

**Comment by fballiano** [comment]
In some cases I've seen the same behavior (empty cron_schedule table), in that case I was running `php cron.php` instead and (if I remember correctly) that worked

(I know it's a workaround)

**Reply by addison74** [reply]
In production (Debian 11 + PHP 7.4), if I truncate the cron_schedule table, the cronjob will run at a moment and fill in place. Periodically there are updates in the table. In the case of DDEV, the behavior is random, I truncate the table, if I run the command ```ddev php /var/www/html/cron.php```, either the table is populated or not. I also took into account the fact that running the cron.sh script in Docker containers could be a problem. I also analyze the configuration of the OpenMage cron as it was done in a closed PR these days, or directly in the configuration file of DDEV using hooks.

Note: The discussion should not take anyone's time, I just wanted to make known the issue and that I am trying to find a solution. DDEV is extraordinary because I no longer run anything in Linux, except the pre-production version. I can have a DDEV instance in less than 2 minutes, with the Magento sample package. In addition, running in WSL is as fast as in Linux, I can switch between PHP versions in 1 minute. I am using PHPStorm in Windows because there are other programs in which I work daily that do not have a counterpart in Linux.

**Reply by fballiano** [reply]
mmmm dunno if it's actually a real thing or not, but I've always seen that cron fills the cron_schedule table once every 2 runs, like the first one fills the table, the next one runs what it has to run, or something like that. but I've never looked into it to see if it's actually true or not.
